### PR TITLE
Add ntlm authentication support for mail

### PR DESCRIPTION
**Pull request by YT315**
Add ntlm authentication support for mail
use "github.com/Azure/go-ntlmssp"

### CODE REVIEW

File: services/mailer/mailer.go [hunk 1]
```diff
@@ -145,6 +146,35 @@ func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
...
+func (a *ntlmAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
+	negotiateMessage, err := ntlmssp.NewNegotiateMessage(a.domain, "")
+	return "NTLM", negotiateMessage, err
+}
+
+// Next next step of SMTP ntlm auth
+func (a *ntlmAuth) Next(fromServer []byte, more bool) ([]byte, error) {
+	if more {
+		if len(fromServer) == 0 {
+			return nil, fmt.Errorf("ChallengeMessage is empty. ")
```
Comment (silverwind) [inline services/mailer/mailer.go]:
```suggestion
			return nil, fmt.Errorf("NTLM ChallengeMessage is empty.")
```

File: services/mailer/mailer.go [hunk 2]
```diff
@@ -145,6 +146,35 @@ func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
...
+	domainNeeded               bool
+}
+
+// NtlmAuth SMTP AUTH NTLM Auth Handler
+func NtlmAuth(username, password string) smtp.Auth {
+	user, domain, domainNeeded := ntlmssp.GetDomain(username)
+	return &ntlmAuth{user, password, domain, domainNeeded}
+}
+
+// Start start SMTP ntlm auth
```
Comment (silverwind) [inline services/mailer/mailer.go]:
```suggestion
// Start starts SMTP NTLM Auth
```

**Review by silverwind — APPROVED** [review]
Can not easily test this, but it seems benign enough

File: services/mailer/mailer.go [hunk 3]
```diff
@@ -145,6 +146,35 @@ func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
...
+func (a *ntlmAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
+	negotiateMessage, err := ntlmssp.NewNegotiateMessage(a.domain, "")
+	return "NTLM", negotiateMessage, err
+}
+
+// Next next step of SMTP ntlm auth
+func (a *ntlmAuth) Next(fromServer []byte, more bool) ([]byte, error) {
+	if more {
+		if len(fromServer) == 0 {
+			return nil, fmt.Errorf("NTLM ChallengeMessage is empty.")
```
Comment (silverwind) [inline services/mailer/mailer.go]:
```suggestion
			return nil, fmt.Errorf("ntlm ChallengeMessage is empty.")
```

**Review by wxiaoguang — APPROVED** [review]

File: services/mailer/mailer.go [hunk 4]
```diff
@@ -145,6 +146,35 @@ func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
...
+func (a *ntlmAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
+	negotiateMessage, err := ntlmssp.NewNegotiateMessage(a.domain, "")
+	return "NTLM", negotiateMessage, err
+}
+
+// Next next step of SMTP ntlm auth
+func (a *ntlmAuth) Next(fromServer []byte, more bool) ([]byte, error) {
+	if more {
+		if len(fromServer) == 0 {
+			return nil, fmt.Errorf("ntlm ChallengeMessage is empty.")
```
Comment (silverwind) [inline services/mailer/mailer.go]:
```suggestion
			return nil, fmt.Errorf("ntlm ChallengeMessage is empty")
```

**Pull request merged**